File: twidere/src/main/java/org/mariotaku/twidere/model/tab/impl/SearchTabConfiguration.java
package org.mariotaku.twidere.model.tab.impl;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import org.mariotaku.twidere.R;
import org.mariotaku.twidere.fragment.StatusesSearchFragment;
import org.mariotaku.twidere.model.Tab;
import org.mariotaku.twidere.model.tab.DrawableHolder;
import org.mariotaku.twidere.model.tab.StringHolder;
import org.mariotaku.twidere.model.tab.TabConfiguration;
import org.mariotaku.twidere.model.tab.argument.TextQueryArguments;
import org.mariotaku.twidere.model.tab.conf.StringExtraConfiguration;

import static org.mariotaku.twidere.constant.IntentConstants.EXTRA_QUERY;

/**
 * Created by mariotaku on 2016/11/27.
 */

public class SearchTabConfiguration extends TabConfiguration {
    @NonNull
    @Override
    public StringHolder getName() {
        return StringHolder.resource(R.string.search);
    }

    @NonNull
    @Override
    public DrawableHolder getIcon() {
        return DrawableHolder.Builtin.SEARCH;
    }

    @AccountFlags
    @Override
    public int getAccountFlags() {
        return FLAG_HAS_ACCOUNT | FLAG_ACCOUNT_REQUIRED;
    }

    @Nullable
    @Override
    public ExtraConfiguration[] getExtraConfigurations(Context context) {
        return new ExtraConfiguration[]{
                new StringExtraConfiguration(EXTRA_QUERY, null).maxLines(1).title(R.string.search_statuses).headerTitle(R.string.query)
        };
    }

    @Override
    public boolean applyExtraConfigurationTo(@NonNull Tab tab, @NonNull ExtraConfiguration extraConf) {
        final TextQueryArguments arguments = (TextQueryArguments) tab.getArguments();
        assert arguments != null;
        switch (extraConf.getKey()) {
            case EXTRA_QUERY: {
                final String query = ((StringExtraConfiguration) extraConf).getValue();
                if (query == null) return false;
                arguments.setQuery(query);
                break;
            }
        }
        return true;
    }

    @Override
    public boolean readExtraConfigurationFrom(@NonNull Tab tab, @NonNull ExtraConfiguration extraConf) {
        final TextQueryArguments arguments = (TextQueryArguments) tab.getArguments();
        if (arguments == null) return false;
        switch (extraConf.getKey()) {
            case EXTRA_QUERY: {
                ((StringExtraConfiguration) extraConf).setValue(arguments.getQuery());
                break;
            }
        }
        return true;
    }

    @NonNull
    @Override
    public Class<? extends Fragment> getFragmentClass() {
        return StatusesSearchFragment.class;
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/model/tab/impl/SearchTabConfiguration.kt
/*
 *             Twidere - Twitter client for Android
 *
 *  Copyright (C) 2012-2017 Mariotaku Lee <mariotaku.lee@gmail.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.model.tab.impl

import android.content.Context
import org.mariotaku.twidere.R
import org.mariotaku.twidere.annotation.TabAccountFlags
import org.mariotaku.twidere.constant.IntentConstants.EXTRA_QUERY
import org.mariotaku.twidere.fragment.statuses.StatusesSearchFragment
import org.mariotaku.twidere.model.Tab
import org.mariotaku.twidere.model.tab.DrawableHolder
import org.mariotaku.twidere.model.tab.StringHolder
import org.mariotaku.twidere.model.tab.TabConfiguration
import org.mariotaku.twidere.model.tab.argument.TextQueryArguments
import org.mariotaku.twidere.model.tab.conf.StringExtraConfiguration

/**
 * Created by mariotaku on 2016/11/27.
 */

class SearchTabConfiguration : TabConfiguration() {

    override val name = StringHolder.resource(R.string.action_search)

    override val icon = DrawableHolder.Builtin.SEARCH

    override val accountFlags = TabAccountFlags.FLAG_HAS_ACCOUNT or TabAccountFlags.FLAG_ACCOUNT_REQUIRED

    override val fragmentClass = StatusesSearchFragment::class.java

    override fun getExtraConfigurations(context: Context) = arrayOf(
            StringExtraConfiguration(EXTRA_QUERY, R.string.search_statuses, null).maxLines(1).headerTitle(R.string.query)
    )

    override fun applyExtraConfigurationTo(tab: Tab, extraConf: ExtraConfiguration): Boolean {
        val arguments = tab.arguments as TextQueryArguments
        when (extraConf.key) {
            EXTRA_QUERY -> {
                val query = (extraConf as StringExtraConfiguration).value?.takeIf(String::isNotBlank) ?: return false
                arguments.query = query
            }
        }
        return true
    }

    override fun readExtraConfigurationFrom(tab: Tab, extraConf: ExtraConfiguration): Boolean {
        val arguments = tab.arguments as? TextQueryArguments ?: return false
        when (extraConf.key) {
            EXTRA_QUERY -> {
                (extraConf as StringExtraConfiguration).value = arguments.query
            }
        }
        return true
    }
}
